pydantic/fields.py

```diff
@@ -411,18 +415,33 @@ def _type_analysis(self) -> None:  # noqa: C901 (ignore complexity)
         elif is_literal_type(self.type_):
             return
 
-        origin = getattr(self.type_, '__origin__', None)
+        origin = get_origin(self.type_)
         if origin is None:
             # field is not "typing" object eg. Union, Dict, List etc.
-            # allow None for virtual superclasses of NoneType, e.g. Hashable
-            if isinstance(self.type_, type) and isinstance(None, self.type_):
-                self.allow_none = True
-            return
+            # It could be a subclass of a "typing" object
+            try:
+                if issubclass(self.type_, list):
+                    while not get_args(self.type_) and hasattr(self.type_, '__orig_bases__'):
+                        self.type_ = self.type_.__orig_bases__[0]
+                    origin = get_origin(self.type_)
+                elif issubclass(self.type_, set):
+                    while not get_args(self.type_) and hasattr(self.type_, '__orig_bases__'):
+                        self.type_ = self.type_.__orig_bases__[0]
+                    origin = get_origin(self.type_)
```

Comment:
This seems to be the same thing again. Any reason why not just `if issubclass(self.type_, (list, set))`?